x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/ExecuteStepsUpdateTask.java

```diff
@@ -169,13 +176,23 @@ public ClusterState execute(final ClusterState currentState) throws IOException
     public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
         if (oldState.equals(newState) == false) {
             IndexMetaData indexMetaData = newState.metaData().index(index);
-            if (nextStepKey != null && nextStepKey != TerminalPolicyStep.KEY && indexMetaData != null) {
-                logger.trace("[{}] step sequence starting with {} has completed, running next step {} if it is an async action",
-                    index.getName(), startStep.getKey(), nextStepKey);
-                // After the cluster state has been processed and we have moved
-                // to a new step, we need to conditionally execute the step iff
-                // it is an `AsyncAction` so that it is executed exactly once.
-                lifecycleRunner.maybeRunAsyncAction(newState, indexMetaData, policy, nextStepKey);
+            if (indexMetaData != null) {
+
+                LifecycleExecutionState exState = LifecycleExecutionState.fromIndexMetadata(indexMetaData);
+                if (ErrorStep.NAME.equals(exState.getStep()) && this.failure != null) {
+                    lifecycleRunner.registerFailedOperation(indexMetaData, failure);
+                } else {
+                    lifecycleRunner.registerSuccessfulOperation(indexMetaData);
+                }
+
+                if (nextStepKey != null && nextStepKey != TerminalPolicyStep.KEY) {
+                    logger.trace("[{}] step sequence starting with {} has completed, running next step {} if it is an async action",
```

Comment:
I don't think we need to here since it is not an expensive trace call (just getters really) and this is not in any kind of loop where performance would be critical